more in-depth quick start guide (#1196)
Azure/azure-service-operator: docs/howto/quickstart.md

# Quick start application deployment using ASO

In this guide we will:
 - Create a Service Principal
 - Create an Azure Container Registry
 - Create an AKS cluster with the ACR attached
 - Deploy ASO
 - Deploy an Azure Votes app that requires Azure SQL Server using ASO

## Set up the Kubernetes environment and deploy ASO

 ### Requirements
- az cli ([install instructinos](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest))
- kubectl cli ([install instructions](https://kubernetes.io/docs/tasks/tools/install-kubectl/))
- helm cli ([install instructions](https://helm.sh/docs/intro/install/))
- jq [install instructions](https://stedolan.github.io/jq/download/)

### Prepare credentials and variables

Set these variables now as they will help with copy/paste of commands later.

>Only the first 3 environment variables need to be changed before pasting.

```
VOTES_APP_INSTANCE_NAME="<alias>" # only alpha numeric
TENANT_ID=""

> SUBSCRIPTION_ID=""

Reviewer: Can we add instructions on how people can get the values for these variables?

Reviewer 2: Please also add some links that describe what those `ID` are

Author: can do!

Author: I'm adding the instructions for finding them. I'm not sure a quick start for someone completely unfamiliar to Azure is entirely in scope here. Plus, I don't know if MS has a definition listed anywhere.

> SUBSCRIPTION_ID=""

Author: hmm, yes, this looks better to me

> SUBSCRIPTION_ID=""

Author: although....not sure if the windows folks will get this to work in powershell...

> SUBSCRIPTION_ID=""

Author: guess since we are already using jq it doesn't matter too much

> SUBSCRIPTION_ID=""

Reviewer 3: We could use `--query` and `--output tsv` instead of jq, similar to the other examples in the file:

```
TENANT_ID=$(az account show --query homeTenantId -o tsv)
SUBSCRIPTION_ID=$(az account show --query id -o tsv)
```

>If you already have a Container registry you can skip this step.

>This guide attaches ACR directly to the AKS cluster. If you use your own Container Registry you may need to create an image pull secret.

```
az acr create \
    --name $ACR_NAME \
    --resource-group $RESOURCE_GROUP \
    --location $LOCATION \
    --sku Basic
```

### Create a Kubernetes cluster
>If you already have a Kubernetes cluster you can skip this step. Any Kubernetes cluster v1.13+ should work here.

#### Create an AKS Cluster attached to the ACR

>You may need to add the `--generate-ssh-keys` flag if you get an error stating the following: `An RSA key file or key value must be supplied to SSH Key Value`
```
az aks create \
    --location $LOCATION \
    --name $AKS_NAME \
    --resource-group $RESOURCE_GROUP \
    --node-count 2 \
    --node-vm-size "Standard_DS2_v2" \
    --attach-acr $ACR_NAME # remove this flag if using your own container registry
```

#### Configure local kubectl to point to new cluster
```
az aks get-credentials \
    --name $AKS_NAME \
    --resource-group $RESOURCE_GROUP
```

### Create Service Principal for Azure Service Operator

> >The access granted for the purposes of this guide may not be appropriate for production deployments

Reviewer: ```suggestion
>The access granted for the purposes of this guide may not be appropriate for production deployments. We'll store the credentials in a JSON file to access later when deploying ASO.
```

```
az ad sp create-for-rbac \
    --sdk-auth \
    --scope "/subscriptions/"$SUBSCRIPTION_ID \
    --role Contributor \
    --name $OPERATOR_SP_NAME > $CREDENTIAL_DIR/operator_sp_creds.json
```

### Install Cert Manager

Cert Manager is used to provide certificates for Kube RBAC Proxy and the operator's web hooks.

```

> kubectl create namespace cert-manager
> kubectl label namespace cert-manager cert-manager.io/disable-validation=true
> kubectl apply --validate=false -f https://github.com/jetstack/cert-manager/releases/download/v0.12.0/cert-manager.yaml

Reviewer: ```suggestion
kubectl apply-f https://github.com/jetstack/cert-manager/releases/download/v0.12.0/cert-manager.yaml
```

REVISION: 1
TEST SUITE: None
NOTES:
Demo App Deployed!

Run `kubectl port-forward deployment/azure-votes-sql 8080:8080`
Then visit `http://localhost:8080` in your browser to test.
```

This command will deploy 5 kinds of Kube resoruces

- ResourceGroup
- AzureSqlServer
- AzureSqlDatabase
- AzureSqlFirewallRule
- Deployment

Use `kubectl get ${Kind}` to see the status of that object.

E.g.

```
kubectl get ResourceGroup
NAME                     PROVISIONED   MESSAGE
app-rg-azvotessqldemo1   true          successfully provisioned

kubectl get AzureSqlServer
NAME                  PROVISIONED   MESSAGE
sql-azurevotesdemo   true          successfully provisioned

kubectl get AzureSqlDatabase
NAME                    PROVISIONED   MESSAGE
sqldb-azvotessqldemo1   true          successfully provisioned

kubectl get AzureSqlFirewallRule
NAME                    PROVISIONED   MESSAGE
saso-sql-fwrule         true          successfully provisioned
saso-sql-fwrule-local   true          successfully provisioned
```

Once all the Azure resources are successfully provisioned (ResourceGroup, AzureSqlServer, AzureSqlDatabase, AzureSqlFirewallRule) the Azure Votes App will be able to start.

> >The Azure resources take 3-4 minutes to become ready. The application will attempt to start once the resources are provisioned but this doesn't mean all networking is in fact ready. For the reason the application is biult to handle failures in connection by exiting so Kubernetes can restart it.

Reviewer: ```suggestion
>[!NOTE]
>The Azure resources typically take 3-4 minutes to become ready. The application will attempt to start once the resources are provisioned but this doesn't mean all networking is in fact ready. For this reason, the application is built to handle connection failures by exiting, so Kubernetes can restart it.
```
